src/index.js

```diff
@@ -279,42 +270,45 @@ function walk(exports, includeDefault) {
 function getMain({ options, entry, format }) {
 	const { pkg } = options;
 	const pkgMain = options['pkg-main'];
+	const pkgTypeModule = pkg.type === 'module';
+	let multipleEntries = options.multipleEntries;
 
 	if (!pkgMain) {
 		return options.output;
 	}
 
-	let mainNoExtension = options.output;
-	if (options.multipleEntries) {
+	let defaultOutput = options.output;
+	if (multipleEntries) {
 		let name = entry.match(new RegExp(/([\\/])index/.source + EXTENSION.source))
-			? mainNoExtension
+			? defaultOutput
 			: entry;
-		mainNoExtension = resolve(dirname(mainNoExtension), basename(name));
+		defaultOutput = resolve(dirname(defaultOutput), basename(name));
 	}
-	mainNoExtension = mainNoExtension.replace(EXTENSION, '');
+	const defaultOutputNoExtension = defaultOutput.replace(EXTENSION, '');
 
 	const mainsByFormat = {};
 
-	mainsByFormat.es = replaceName(
-		pkg.module && !pkg.module.match(/src\//)
-			? pkg.module
-			: pkg['jsnext:main'] || 'x.esm.js',
-		mainNoExtension,
+	mainsByFormat.es = resolve(
+		(!multipleEntries &&
+			(pkg.module && !pkg.module.match(/src\//)
+				? pkg.module
+				: pkg['jsnext:main'])) ||
+			`${defaultOutputNoExtension}.esm.js`,
 	);
-	mainsByFormat.modern = replaceName(
-		(pkg.exports && walk(pkg.exports, pkg.type === 'module')) ||
-			(pkg.syntax && pkg.syntax.esmodules) ||
-			pkg.esmodule ||
-			'x.modern.js',
-		mainNoExtension,
+	mainsByFormat.modern = resolve(
+		(!multipleEntries &&
+			((pkg.exports && walk(pkg.exports, pkgTypeModule)) ||
+				(pkg.syntax && pkg.syntax.esmodules) ||
+				pkg.esmodule)) ||
+			`${defaultOutputNoExtension}.modern.${pkgTypeModule ? 'js' : 'mjs'}`,
```

Comment:
Missed this in #802: if the CJS output can respond to package type, so should the modern output I believe.